### PR TITLE
Allow passing into the crypto key via ASTRO_KEY

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- This makes it possible to pass in the cryptography used via the ASTRO_KEY environment variable.
- This allows you to reuse the key in each build, ensuring that if you have separate server and client deployments that the keys are never out of sync.
- Fixes https://github.com/withastro/astro/issues/11851

- Keys are generated using `astro create-key` command.

## Testing

- New E2E test

## Docs

N/A

### CODE REVIEW

**Comment by matthewp** [comment]
!preview crypto-env

**Comment by sasoria** [comment]
I've tested this at least five times by generating an AES-256 key and adding this to package.json:
```js
"build": ASTRO_KEY=MY_KEY astro build
```

The application was then built with,
```js
npm run build
```

I did not see any of the errors I've seen before during deploys, so this should fix the issue I had on rolling updates.  Thank you for fixing this so quickly.

**Comment by matthewp** [comment]
Thanks @sasoria, I still need to write a test so this probably won't go out today.

File: packages/astro/src/core/build/index.ts [hunk 1]
```diff
@@ -230,6 +233,13 @@ class AstroBuilder {
 			cacheManifest: internals.cacheManifestUsed,
 		});
 
+		if(!hasKey && opts.settings.config.experimental.serverIslands) {
+			this.logger.info('build', `This build generated a key to encrypt props passed to Server islands. To reuse the same key across builds, set this value as ASTRO_KEY in an environment variable on your build server.
+
+ASTRO_KEY=${await encodeKey(await keyPromise)}
```
Comment (ascorbic) [inline packages/astro/src/core/build/index.ts:239]: Is it safe to print this? I'd be worried it would show in public logs

Comment (matthewp) [reply packages/astro/src/core/build/index.ts:239]: Ah yeah good point. Hm, I'm not sure what to do then. The only other way I can think to do this is with a new command. Something like `astro generate-key` or something.

Comment (matthewp) [reply packages/astro/src/core/build/index.ts:239]: That's probably a better way to go. Will update the PR.

Comment (matthewp) [reply packages/astro/src/core/build/index.ts:239]: New output is like:

```
Generated a key to encrypt props passed to Server islands. To reuse the same key across builds, set this value as ASTRO_KEY in an environment variable on your build server.

ASTRO_KEY=P90X3r0+nEqzystDC1pg01VS4s/+jINyDTYSNlEO0HQ=
```

**Review by ascorbic — APPROVED** [review]

**Pull request merged**